### PR TITLE
Extension is slow to activate #4226

### PR DESCRIPTION
**Pull request by vrubezhny**
The PR adds some optimization to the extension activation.

Also, the PR removes function 'migrateFromOdo018' which looks over outdated and doesn't make its work properly after a numerous changes in ODO. See: #4255

Fixes: #4255
Fixes: #4226

### CODE REVIEW

**Comment by vrubezhny** [comment]
My calculations for the extension loading time before (installed from MS Marketplace) and after the fix (installed from locally built VSIX) differ in numbers, here are some examples:

- Before the fix (installed from MS Marketplace):
Screenshot from 2024-07-03 22-45-44

- After the fix (installed from locally built VSIX):
Screenshot from 2024-07-03 22-40-32

File: src/extension.ts [hunk 1]
```diff
@@ -258,34 +244,38 @@ export async function activate(extensionContext: ExtensionContext): Promise<unkn
         }
     }
 
-    OpenShiftExplorer.getInstance().onDidChangeContextEmitter.event(async (context) => {
-        const kind = await getNamespaceKind();
-        void Oc.Instance.getActiveProject().then((namespace) =>
-            updateContextStatusBarItem(activeNamespaceStatusBarItem, 'project-node', namespace, `Current ${kind}: ${namespace}`,
-                !isNamespaceInfoStatusBarDisabled()));
+    const updateContextAndProjectStatusBarItems = async function(context?: string): Promise<void> {
+        return getNamespaceKind().then(async (kind) => {
```
Comment (datho7561) [inline src/extension.ts:248]: this function is async. You can await `getNamespaceKind` instead of using `.then(...`

File: src/extension.ts [hunk 2]
```diff
@@ -72,9 +59,9 @@ async function registerKubernetesCloudProvider(): Promise<void> {
 
 export async function activate(extensionContext: ExtensionContext): Promise<unknown> {
     void WelcomePage.createOrShow();
-    await commands.executeCommand('setContext', 'isVSCode', env.uiKind);
+    void commands.executeCommand('setContext', 'isVSCode', env.uiKind);
     // UIKind.Desktop ==1 & UIKind.Web ==2. These conditions are checked for browser based & electron based IDE.
```
Comment (datho7561) [inline src/extension.ts:63]: Please remove this comment, it doesn't seem relevant anymore

**Review by datho7561 — CHANGES_REQUESTED** [review]
Seems pretty good so far. Noticeably speeds up the extension start up. I think the main speed up is not blocking on registering the commands/webviews on startup. I was worried this would introduce concurrency bugs, but it doesn't seem to introduce any.

I've added two small comments.

**Comment by vrubezhny** [comment]
There's still something strange in the extension activation time calculation... 


The extension's `activate` method

File: src/extension.ts (hunk 1 above)
Comment (vrubezhny) [reply src/extension.ts]: ACK

File: src/extension.ts (hunk 2 above)
Comment (vrubezhny) [reply src/extension.ts]: Removed

**Comment by vrubezhny** [comment]
> Seems pretty good so far. Noticeably speeds up the extension start up. I think the main speed up is not blocking on registering the commands/webviews on startup. I was worried this would introduce concurrency bugs, but it doesn't seem to introduce any.

Still trying to introduce anything that breaks it :sunglasses: ... I've turned setting up the Kube Context call to be async (it takes quite a lot of time to finish and also it might require a user intervention, so...) and wait at the end for it's promise to be resolved... Still doesn't look like I broke anything (even the status bar active context and project items set up).

@datho7561 Anyway, take a look please

> I've added two small comments.

Done

**Comment by datho7561** [comment]
Try out this patch to do the timing:

```diff-typescript
diff --git a/src/extension.ts b/src/extension.ts
index 62cdd465..4a39e36d 100644
--- a/src/extension.ts
+++ b/src/extension.ts
@@ -3,6 +3,8 @@
  *  Licensed under the MIT License. See LICENSE file in the project root for license information.
  *-----------------------------------------------------------------------------------------------*/
 
+const startTime = new Date().getTime();
+import { Context as KcuContext } from '@kubernetes/client-node/dist/config_types';
 import {
     authentication,
     commands, env, ExtensionContext, languages, QuickPickItemKind,
@@ -25,16 +27,15 @@ import { startTelemetry } from './telemetry';
 import { ToolsConfig } from './tools';
 import { TokenStore } from './util/credentialManager';
 import { getNamespaceKind, KubeConfigUtils, setKubeConfig } from './util/kubeUtils';
-import { Context as KcuContext } from '@kubernetes/client-node/dist/config_types';
 import { setupWorkspaceDevfileContext } from './util/workspace';
 import { registerCommands } from './vscommand';
 import { OpenShiftTerminalManager } from './webview/openshift-terminal/openShiftTerminal';
 import { WelcomePage } from './welcomePage';
 import { registerYamlHandlers } from './yaml/yamlDocumentFeatures';
 
-import { Oc } from './oc/ocWrapper';
-import { K8S_RESOURCE_SCHEME, K8S_RESOURCE_SCHEME_READONLY, KubernetesResourceVirtualFileSystemProvider } from './k8s/vfs/kuberesources.virtualfs';
 import { KubernetesResourceLinkProvider } from './k8s/vfs/kuberesources.linkprovider';
+import { K8S_RESOURCE_SCHEME, K8S_RESOURCE_SCHEME_READONLY, KubernetesResourceVirtualFileSystemProvider } from './k8s/vfs/kuberesources.virtualfs';
+import { Oc } from './oc/ocWrapper';
 
 // eslint-disable-next-line @typescript-eslint/no-empty-function
 // this method is called when your extension is deactivated
@@ -276,6 +277,9 @@ export async function activate(extensionContext: ExtensionContext): Promise<unkn
     // Wait for finishing Kube Config setup
     await setKubeConfigPromise;
 
+    // eslint-disable-next-line no-console
+    console.log(`vscode-openshift activation time: ${new Date().getTime() - startTime}`);
+
     return {
         verifyBundledBinaries
     };

```

**Comment by vrubezhny** [comment]
> Try out this patch to do the timing:

Yep... 
image

Now the reported activation times are closer to each other... even the vscode's time is less than the Extension activation time (?)...

But I cannot do anything to speed up the time spent to import the required dependencies...

**Comment by datho7561** [comment]
> But I cannot do anything to speed up the time spent to import the required dependencies... Can I?

Not really. *Maybe* we could all the code into one file, but we would need to make sure that we don't use any libraries that prevent that, and we can investigate/do that in a future PR.

I'll take another look at the changes and make sure it looks good.

**Review by datho7561 — APPROVED** [review]
This does the trick for me. I no longer have to wait a few seconds to interact with the Application Explorer. It loads in what feels like around a second.


I don't know if you want to wait on feedback from Fred, or go ahead and merge. Either is fine to me.

**Comment by vrubezhny** [comment]
> Maybe we could all the code into one file, but we would need to make sure that we don't use any libraries that prevent that, and we can investigate/do that in a future PR.

Yeah. I got the idea.

@datho7561 Let's merge this one. And then we can ask Fred for his feedback and improve it later.

**Pull request merged**